src/workerd/api/sql-test.js

```diff
@@ -0,0 +1,81 @@
+function assertEqual(a, b) {
+  if (a !== b) {
+    throw new Error(a + " !== " + b);
+  }
+}
+
+function requireException(callback, expectStr) {
+  try {
+    callback();
+  } catch (err) {
+    const errStr = `${err}`;
+    if (!errStr.includes(expectStr)) {
+      throw new Error(`Got unexpected exception '${errStr}', expected: '${expectStr}'`);
+    }
+    return;
+  }
+  throw new Error(`Expected exception '${expectStr}' but none was thrown`);
+}
+
+export default {
+  async test(ctrl, env, ctx) {
+
+    const sql = new SqlDatabase();
+
+    // Test numeric results
+    const resultNumber = sql.exec("SELECT 123", null).results;
+    assertEqual(resultNumber.length, 1);
+    assertEqual(resultNumber[0]["123"], 123);
```

Comment:
I definitely agree with @kentonv that an iterator based API would be better (and safer here). A sync-iterator if necessary, but async-iterator if possible... e.g. 

```js
for (const row of sql.exec("SELECT 123", null) {
  console.log(row.length);
  console.log(row[0]);
}

// or 
for await (const row of sql.exec("SELECT 123", null) {
  console.log(row.length);
  console.log(row[0]);
}
```

Another possibility to think about later (but very optional right now)... would be a stream-based consumer...

```js
const { stream } = sql.exec("SELECT 123", null);
const reader = stream.getReader();
const row = await reader.read().value;
console.log(row.length);
console.log(row[0]);
```

This would have the advantage of allowing streaming transformations of result values using standard APIs... e.g.

```js
const rowTransform = new TransformStream({
  transform(chunk, c) {
    // transform row...
  }
});

sql.exec("SELECT 123", null).pipeThrough(rowTransform)
```

That said, I would absolutely start with just having the iteratable API that @kentonv suggests.

Reply:
oh, I'd also point out that none of these options are mutually exclusive. We can start with sync iterator and add the others later.

Reply:
I think we're settled on providing a sync interface. To provide async we'd need to use fibers, since SQLite itself is sync. But that'd be a lot of plumbing which would likely just perform worse in the end. IMO file I/O should almost always be sync.